### ISSUE TITLE
Skolem type in same block as term causes assertion violation.

### ISSUE THREAD

**Issue by MichaelRawson** [vprover/vampire#421]
The input

```
tff(test, conjecture, ![A: $tType]: ![X: A]: X = X).
```

causes the following assertion violation:

```
Condition in file /home/michael/vampire/Shell/Skolem.cpp, line 564 violated:
first_pass || sym == last_sym+1
----- stack dump -----
Version : Vampire 4.7 (commit 38d54c2ae on 2022-10-12 12:12:18 +0100)
Control points passed: 35324
last control point:
          -Substitution::bind(int,Term*) (35324)
main
 vampireMode()
  doProving()
   getPreprocessedProblem
    Preprocess::preprocess
     Preprocess::preprocess3(Problem&)
      Preprocess::preprocess3(Unit*)
       Skolem::skolemise(Unit*)
        Skolem::skolemiseImpl(FormulaUnit*)
         Skolem::skolemise (Formula*)
----- end of stack dump -----
```

This is because in `Skolem.cpp` at the indicated line, we assert that fresh Skolem symbols have sequential indices. However, the Skolem for `A` is a type constructor, not a function symbol, so we fail here. @quickbeam123 - if you get time to look at this, is there an easy way to fix this, or is the technique doomed?

**Comment by ibnyusuf** [comment]
I came across this issue on a branch I am working on, and changed the assertion to:

`ASS(!name_reuse || first_pass || sym == last_sym+1);`

which seemed to fix thigns. Not sure if this is valid though...

**Comment by MichaelRawson** [comment]
> Could it be that the code is fine, but the assertion is invalid?

It's certainly broken _with_ name reuse on. Without it the code is fine (I think) but the assertion is too aggressive.

**Comment by quickbeam123** [comment]
Good points! I am processing this now :)

**Comment by quickbeam123** [comment]
OK, so this would be fine if the two quantifiers stayed in two distinct blocks. But flattening comes and merges the two blocks togeter. However, modifying flattening is not the right fix, since the user can already write them merged. Should we split blocks which mix standard sorts and `$tType`? Or should we simply write a more expensive version of symbol reuse in Skolem which does not use the "consecutiveness" trick? Or should we admit symbol reuse does not help that much but complicates the code considerably and remove it again?

By the way, I was surprised to see this
`tff(test, conjecture, ![X: A] : ![A: $tType] : X = X).`
parses (notice that `A`, when it first occurs, has not been declared yet.)
Actuall, even this parses:
`tff(test, conjecture, ![X: A] : X=X).`
and `A` is assumed to be universally quantified something there.

**Comment by MichaelRawson** [comment]
>  Or should we admit symbol reuse does not help that much but complicates the code considerably and remove it again?

I'm inclined towards this. I thought it would be a quick and relatively easy extension, orthogonal to the rest of Vampire, that also helped with the thread-parallel Vampire project. It turns out that none of these are really true! I'm glad that we did the work (the PAAR paper was interesting, at least to me), but it seems like a source of bugs for no real gain. How do you feel about it?

IIRC at some point we want to have formula sharing, at which point this becomes easier.

> even this parses:

TFF1 spec says "All type variables must be bound by a !>-binder.", so we'd be within our rights to not parse that. Do we want to be generous and universally-quantify unbound variables? We do currently, even for FOF I think.

**Comment by quickbeam123** [comment]
I think that scheduling name reuse for removal is the grown-up way of proceeding. Hopefully, it will not be too difficult to do the surgery. (And yes, in the distant future, formula sharing should bring the same benefits and be more principled at the same time.)

**Comment by MichaelRawson** [comment]
OK, then I will schedule it for deletion. In a similar vein do you have data from Snake to suggest the removal of other less-useful options?

**Comment by quickbeam123** [comment]
I have tons of data from Snake and didn't find time to look at it yet. Please keep asking me about as this could really be interesting and worthwhile!

**Comment by quickbeam123** [comment]
A related issue is that we currently decided not to shuffle quantifiers under polymorphism as such creatures can appear in the input:
`! [B: $tType,A: $tType,G: fun(A,B),F: fun(A,B)] :`
It was suggested that rectification (does it only come after flattening?) should always split these into type quants and fo quants. Then we could shuffle each separately, name reuse in skolem could use its trick and all would be nice.

This even tempts me to consider having a type-var quantifier a separate creature (to standard FORALL and EXISTS) and if the input language (TPTP) does not support this view of separating them, maybe vampire internally could (to get rid of issues like this one in a principled way).

**Comment by MichaelRawson** [comment]
Fixed by #444, albeit by removing the buggy feature.
